bin/picom-trans

```diff
@@ -160,95 +160,93 @@ parse_args()
 } #}}
 
 
-# NOTE: The code inside the following function has been taken directly from the
-# old version of picom-trans.  Tread cautiously!
 get_target_window_id()
 { #{{
 
-  # Get ID of the target window
-  if test -z "$winidtype"; then
-    treeout=$(xwininfo -children -frame)
-  else
-    test "$winidtype" = "-id" \
-      && ! echo "$winid" | grep -Eiq '^[[:space:]]*(0x[[:xdigit:]]+|[[:digit:]]+)[[:space:]]*$' \
-      && echo "Bad window ID." >&2 \
-      && exit 1
-    treeout=$(xwininfo -children $winidtype "$winid")
+  # Get the output of xwininfo
+  if test -z "$winidtype"
+  then xwininfo_output="$(xwininfo -children -frame)"
+  elif test "$winidtype" = "-name"
+  then xwininfo_output="$(xwininfo -children -name "$winid")"
+  elif test "$winidtype" = "-id"
+  then
+    # First, check if supplied window id is valid
+    if ! echo "$winid" | grep -Eiq '^[[:space:]]*(0x[[:xdigit:]]+|[[:digit:]]+)[[:space:]]*$'
+    then
+      echo "Bad window ID" >&2
+      exit 1
+    fi
+    xwininfo_output="$(xwininfo -children -id "$winid")"
   fi
 
-  wid=$(echo "$treeout" | sed -n 's/^xwininfo:.*: \(0x[[:xdigit:]]*\).*$/\1/p')
-
-  if test -z "$wid"; then
-    echo "Failed to find window." >&2
+  # Extract window id from xwininfo output
+  winid="$(echo "$xwininfo_output" | sed -n 's/^xwininfo:.*: \(0x[[:xdigit:]]*\).*$/\1/p')"
+  if test -z "$winid"
+  then
+    echo "Failed to find window" >&2
     exit 1
   fi
 
   # Make sure it's not root window
-  if echo "$treeout" | grep -Fq "Parent window id: 0x0"; then
-    echo "Cannot set opacity on root window." >&2
+  if echo "$xwininfo_output" | grep -Fq "Parent window id: 0x0"
+  then
+    echo "Cannot set opacity on root window" >&2
     exit 1
   fi
 
-  # If it's already the topmost window
-  if echo "$treeout" | grep -q 'Parent window id: 0x[[:xdigit:]]* (the root window)'; then
-    topmost=$wid
-  else
-    # Get the whole window tree
-    treeout=$(xwininfo -root -tree)
-
-    if test -z "$treeout"; then
-      echo "Failed to get root window tree." >&2
+  # If it's not the topmost window, get the topmost window
+  if ! echo "$xwininfo_output" | grep -q 'Parent window id: 0x[[:xdigit:]]* (the root window)'
+  then
+    window_tree="$(xwininfo -root -tree)"
+    if test -z "$window_tree"
+    then
+      echo "Failed to get root window tree" >&2
       exit 1
     fi
 
-    # Find the line number of the target window in the window tree
-    lineno=$(printf "%s" "$treeout" | grep -nw "^\s*$wid" | head -n1 | cut -d ':' -f 1)
-
-    if test -z "$lineno"; then
-      echo "Failed to find window in window tree." >&2
+    # Find the highest ancestor of the target window
+    winid="$(echo "$window_tree" \
+      | sed "/^\s*$wid/q;s/^     \(0x[[:xdigit:]]*\).*/\1/p" \
+      | tail -n 1)"
+    if test -z "$winid"
+    then
+      echo "Failed to find window in window tree" >&2
       exit 1
     fi
```

Comment:
This did break the "find the highest ancestor" functionality. The variable is called `$winid` and you have to silence `sed` with `-n`.

Reply:
:facepalm:

Reply:
72b55db